File: android/src/main/java/xyz/luan/audioplayers/ByteDataSource.java
package xyz.luan.audioplayers;

import android.media.MediaDataSource;
import android.os.Build;

import java.io.*;

public class ByteDataSource extends MediaDataSource {
    private byte[] data;

    public ByteDataSource(byte[] realData) {
        data = realData;
    }

    @Override
    public synchronized long getSize() throws IOException {
        return data == null ? 0 : data.length;
    } 

    @Override
    public synchronized void close() throws IOException {
    }

    @Override
    public synchronized int readAt(long position, byte[] buffer, int offset, int size) {
        if (position >= data.length) {
            return -1;
        }

        if (position + size > data.length) {
            size -= ((int)position + size) - data.length;
        }
        
        System.arraycopy(data, (int)position, buffer, offset, size);
        return size;
    }
}

File: android/src/main/kotlin/xyz/luan/audioplayers/ByteDataSource.kt
package xyz.luan.audioplayers

import android.media.MediaDataSource
import java.io.IOException
import kotlin.jvm.Throws

class ByteDataSource(
        private val data: ByteArray,
) : MediaDataSource() {
    @Synchronized
    @Throws(IOException::class)
    override fun getSize(): Long {
        return data.size.toLong()
    }

    @Synchronized
    @Throws(IOException::class)
    override fun close() {
    }

    @Synchronized
    override fun readAt(position: Long, buffer: ByteArray, offset: Int, size: Int): Int {
        if (position >= data.size) {
            return -1
        }

        var remainingSize = size
        if (position + remainingSize > data.size) {
            remainingSize -= position.toInt() + remainingSize - data.size
        }
        System.arraycopy(data, position.toInt(), buffer, offset, remainingSize)
        return remainingSize
    }

}